### DISCUSSION TITLE
Socks5/https proxy support for OpenAI API

### DISCUSSION THREAD

**Discussion by Jipok** [open-webui/open-webui#2268] [General]
OpenAI is not available in some countries. Usually I just specify a proxy, but I couldn't find such a setting for either Open-WebUI or LiteLLM.

**Comment by voiprostov** [comment]
Have you found a solution?

**Reply by firengate** [reply]
Installing packages.
pip install requests[socks]

Add environment variables.
https_proxy=socks5://127.0.0.1:1080

**Reply by voiprostov** [reply]
When I try to run it I get the error:

```
root@3230b052e938:/app/backend# pip install requests[socks]
Requirement already satisfied: requests[socks] in /usr/local/lib/python3.11/site-packages (2.32.3)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.4.0)
Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.10)
Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2.2.3)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2024.8.30)
ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.


[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
root@3230b052e938:/app/backend# pip install --upgrade pip
Requirement already satisfied: pip in /usr/local/lib/python3.11/site-packages (24.0)
ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.


[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: pip install --upgrade pip

```

**Reply by yusiwen** [reply]
> When I try to run it I get the error:
> 
> ```
> root@3230b052e938:/app/backend# pip install requests[socks]
> Requirement already satisfied: requests[socks] in /usr/local/lib/python3.11/site-packages (2.32.3)
> Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.4.0)
> Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.10)
> Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2.2.3)
> Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2024.8.30)
> ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.
> 
> 
> [notice] A new release of pip is available: 24.0 -> 24.3.1
> [notice] To update, run: pip install --upgrade pip
> root@3230b052e938:/app/backend# pip install --upgrade pip
> Requirement already satisfied: pip in /usr/local/lib/python3.11/site-packages (24.0)
> ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.
> 
> 
> [notice] A new release of pip is available: 24.0 -> 24.3.1
> [notice] To update, run: pip install --upgrade pip
> ```

If you're using http/https or mixed proxy server(e.g. Clash), you don't need to install additional dependencies, just set the http_proxy and https_proxy environment variables to your http proxy server address. It currently works perfectly for me.

**Reply by firengate** [reply]
> When I try to run it I get the error:
> 
> ```
> root@3230b052e938:/app/backend# pip install requests[socks]
> Requirement already satisfied: requests[socks] in /usr/local/lib/python3.11/site-packages (2.32.3)
> Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.4.0)
> Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (3.10)
> Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2.2.3)
> Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/site-packages (from requests[socks]) (2024.8.30)
> ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.
> 
> 
> [notice] A new release of pip is available: 24.0 -> 24.3.1
> [notice] To update, run: pip install --upgrade pip
> root@3230b052e938:/app/backend# pip install --upgrade pip
> Requirement already satisfied: pip in /usr/local/lib/python3.11/site-packages (24.0)
> ERROR: Could not install packages due to an OSError: Missing dependencies for SOCKS support.
> 
> 
> [notice] A new release of pip is available: 24.0 -> 24.3.1
> [notice] To update, run: pip install --upgrade pip
> ```

A mistake happened. You should first install the dependency and only after that set the proxy in the variable, as it will interfere with pip's operation.

**Comment by steveepreston** [comment]
How to set proxy when using open-webui via Docker?

**Reply by Jekson353** [reply]
It's very easy: 
In docker-compose.yml:

```
services:
 open-webui:
 image: ghcr.io/open-webui/open-webui:latest
 container_name: open-webui
 ports:
 - "3000:8080"
 environment:
 - HTTP_PROXY=http://proxy.local:3128
 - HTTPS_PROXY=http://proxy.local:3128
 - NO_PROXY=localhost,127.0.0.1,::1,host.docker.internal,ollama
```

If proxy requires login/password: http://user:pass@proxy.local:3128.
NO_PROXY should include all local addresses/services that container goes directly to (e.g. ollama, host.docker.internal, host IP, internal domains).

But remember: If you have access to OpenAI Plus, you won't have access to the API. **API access is charged separately!**
